returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,226 @@
+        map_outputs: Optional[Dict[str, Any]] = None,
+        _meta_info_cache: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ):
+        """
+        :param dataset: inner dataset to be post-processed
+        :param map_seq: post processor function operating on the single-segment level.
+        :param map_seq_stream: post processor function operating on the multiple segment level via an iterator.
+            Allows merging multiple segments into one, or generating multiple output segments from one input segment.
+            When both functions are specified, the segment-level function is applied first to every segment, and the
+            results are passed to the iterator-level function.
+        :param map_outputs: Type and axis specification of the outputs of the mapping functions,
+            like extern_data and model_outputs.
+            To simplify the common case when no shapes change, this value can be left unspecified. The dataset then
+            assumes the same data layout as returned by the wrapped dataset.
+        :param _meta_info_cache: internal usage
+        :param kwargs: see :class:`CachedDataset2`, :class:`Dataset`
+        """
+        super().__init__(**kwargs)
+
+        if self.seq_ordering != "default":
+            raise ValueError(f"specify seq_ordering in wrapped dataset, not in {self.__class__.__name__}")
+        if map_seq is None and map_seq_stream is None:
+            raise ValueError("need to either set map_seq or map_seq_stream")
+        if map_seq and map_seq_stream:
+            raise ValueError("cannot set both map_seq and map_seq_stream")
+
+        self._dataset_def = dataset
+        self._map_seq = map_seq
+        self._map_seq_stream = map_seq_stream
+        self._map_outputs = map_outputs
+
+        self._dataset: Optional[Dataset] = None
+        self._data_keys: Optional[List[str]] = None
+        self._data_iter: Optional[Iterator[Tuple[int, TensorDict]]] = None
+        self._default_input: Optional[str] = None
+        self._dim_template_dict: Optional[TensorDict] = None
+
+        if _meta_info_cache:
+            self._default_input = _meta_info_cache["_default_input"]
+            self._dim_template_dict = _meta_info_cache["_dim_template_dict"]
+            self._estimated_num_seqs = _meta_info_cache["_estimated_num_seqs"]
+            self.labels = _meta_info_cache["labels"]
+            self.num_inputs = _meta_info_cache["num_inputs"]
+            self.num_outputs = _meta_info_cache["num_outputs"]
+
+    @property
+    def _meta_info_cache(self) -> Optional[Dict[str, Any]]:
+        if self.num_outputs is None:
+            return None
+        return {
+            "_default_input": self._default_input,
+            "_dim_template_dict": self._dim_template_dict,
+            "_estimated_num_seqs": self._estimated_num_seqs,
+            "labels": self.labels,
+            "num_inputs": self.num_inputs,
+            "num_outputs": self.num_outputs,
+        }
```

Comment:
If you do all the init of the `TensorDict` and thus also `labels`, `num_outputs` etc already in `__init__`, as we discussed, then I think this `_meta_info_cache` is not really needed.